pkg/controller/deployment/deployment_controller.go

```diff
@@ -22,68 +22,311 @@ import (
+	if d := dc.getDeploymentForRC(curRC); d != nil {
+		dc.enqueueDeployment(d)
+	}
+	// A number of things could affect the old deployment: labels changing,
+	// pod template changing, etc.
+	oldRC := old.(*api.ReplicationController)
+	// TODO: Is this the right way to check this, or is checking names sufficient?
+	if !api.Semantic.DeepEqual(oldRC, curRC) {
+		if oldD := dc.getDeploymentForRC(oldRC); oldD != nil {
+			dc.enqueueDeployment(oldD)
+		}
 	}
-	for _, deployment := range list.Items {
-		if err := d.reconcileDeployment(&deployment); err != nil {
-			return fmt.Errorf("error in reconciling deployment: %v", err)
+}
+
+// When a controller is deleted, enqueue the deployment that manages it.
+// obj could be an *api.ReplicationController, or a DeletionFinalStateUnknown
+// marker item.
+func (dc *DeploymentController) deleteRC(obj interface{}) {
+	rc, ok := obj.(*api.ReplicationController)
+
+	// When a delete is dropped, the relist will notice a pod in the store not
+	// in the list, leading to the insertion of a tombstone object which contains
+	// the deleted key/value. Note that this value might be stale. If the RC
+	// changed labels the new deployment will not be woken up till the periodic resync.
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			glog.Errorf("Couldn't get object from tombstone %+v, could take up to %v before a deployment recreates/updates controllers", obj, FullDeploymentResyncPeriod)
+			return
+		}
+		rc, ok = tombstone.Obj.(*api.ReplicationController)
+		if !ok {
+			glog.Errorf("Tombstone contained object that is not an rc %+v, could take up to %v before a deployment recreates/updates controllers", obj, FullDeploymentResyncPeriod)
+			return
 		}
 	}
-	return nil
+	if d := dc.getDeploymentForRC(rc); d != nil {
+		dc.enqueueDeployment(d)
+	}
+}
+
+// obj could be an *api.Deployment, or a DeletionFinalStateUnknown marker item.
+func (dc *DeploymentController) enqueueDeployment(obj interface{}) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %+v: %v", obj, err)
+		return
+	}
+
+	// TODO: Handle overlapping deployments better. Either disallow them at admission time or
+	// deterministically avoid syncing deployments that fight over RC's. Currently, we only
+	// ensure that the same deployment is synced for a given RC. When we periodically relist
+	// all deployments there will still be some RC instability. One way to handle this is
+	// by querying the store for all deployments that this deployment overlaps, as well as all
+	// deployments that overlap this deployments, and sorting them.
+	dc.queue.Add(key)
+}
+
+func (dc *DeploymentController) Run(workers int, stopCh <-chan struct{}) {
+	defer util.HandleCrash()
+	go dc.dController.Run(stopCh)
+	go dc.rcController.Run(stopCh)
+	go dc.podController.Run(stopCh)
+	for i := 0; i < workers; i++ {
+		go util.Until(dc.worker, time.Second, stopCh)
```

Comment:
I worry that the deployment controller is extremely bursty.  I.e., a new
security update comes out and O(all) deployments in the cluster are
updated.  Back of the envelope, N is still in low thousands for the largest
deployments, which means we would be looking at a queue depth of N.  Do we
suffer from head-of-line blocking?

That's another reason we ran as pods on the cluster - we implicitly are
limited by cluster scale.

On Mon, Sep 21, 2015 at 10:38 AM, Dan Mace <notifications@github.com> wrote:

> In pkg/controller/deployment/deployment_controller.go
> <https://github.com/kubernetes/kubernetes/pull/14130#discussion_r39977813>
> :
>
> > +	// TODO: Handle overlapping deployments better. Either disallow them at admission time or
> > +	// deterministically avoid syncing deployments that fight over RC's. Currently, we only
> > +	// ensure that the same deployment is synced for a given RC. When we periodically relist
> > +	// all deployments there will still be some RC instability. One way to handle this is
> > +	// by querying the store for all deployments that this deployment overlaps, as well as all
> > +	// deployments that overlap this deployments, and sorting them.
> > +	dc.queue.Add(key)
> > +}
> > +
> > +func (dc *DeploymentController) Run(workers int, stopCh <-chan struct{}) {
> > +	defer util.HandleCrash()
> > +	go dc.dController.Run(stopCh)
> > +	go dc.rcController.Run(stopCh)
> > +	go dc.podController.Run(stopCh)
> > +	for i := 0; i < workers; i++ {
> > +		go util.Until(dc.worker, time.Second, stopCh)
>
> Scale-up operations (or the forthcoming post scale up readiness waits) for
> a given deployment might block for a very long time. On a busy system with
> a nontrival number of long running deployments, the worker pool could be
> exhausted and throughput will be diminished. To scale these controllers
> would require increasing the worker could for the process. If that wasn't
> enough, the controller would need scaled out horizontally with more
> controller processes. Is that considered an acceptable scaling strategy for
> this component?
>
> To be fair, the worker goroutines are going to be spending the vast
> majority of their time in sleeping state while awaiting conditions
> (hopefully), so perhaps the upper bound of workers per process is higher
> than one might expect.
>
> cc @smarterclayton <https://github.com/smarterclayton>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/14130/files#r39977813>.
>